controllers/providers/aws/aws.go

```diff
@@ -507,3 +533,278 @@ func IsStackInConditionState(key string, condition string) bool {
+		}
+		return err
+	}
+	role := &iam.DeleteRoleInput{
+		RoleName: w.RoleName,
+	}
+	_, err = w.IamClient.DeleteRole(role)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+
+			case iam.ErrCodeNoSuchEntityException:
+				log.Infof("DeleteDefaultRolePolicy - ErrCodeNoSuchEntityException: %v:", aerr.Error())
+				return nil
+			default:
+				log.Errorf("DeleteDefaultRolePolicy - %v:", aerr.Error())
+				return aerr
+			}
+		} else {
+			log.Errorf("DeleteDefaultRolePolicy - %v:", err)
+		}
+	}
+	return err
+}
+func (w *AwsFargateWorker) CreateDefaultRoleName() *string {
+	s := fmt.Sprintf("%v_%v_Role", *w.ClusterName, *w.ProfileName)
+	return &s
+}
+
+func (w *AwsFargateWorker) CreateDefaultRolePolicy() (*string, error) {
+	var roleName = w.CreateDefaultRoleName()
+	var template = `{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"eks-fargate-pods.amazonaws.com"},"Action":"sts:AssumeRole"}]}`
+	role := &iam.CreateRoleInput{
+		AssumeRolePolicyDocument: &template,
+		Path:                     aws.String("/"),
+		RoleName:                 roleName,
+	}
+	createRoleOutput, err := w.IamClient.CreateRole(role)
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == iam.ErrCodeEntityAlreadyExistsException {
+				// Role already exists.  Need the arn
+				getRoleInput := &iam.GetRoleInput{
+					RoleName: roleName,
+				}
+				getRoleOutput, err := w.IamClient.GetRole(getRoleInput)
+				if err != nil {
+					return nil, err
+				}
+				w.RoleName = getRoleOutput.Role.RoleName
+				return getRoleOutput.Role.Arn, nil
+			}
+			return nil, aerr
+		}
+		return nil, err
+	} else {
+		w.RoleName = roleName
+		rolePolicy := &iam.AttachRolePolicyInput{
+			PolicyArn: aws.String(defaultPolicyArn),
+			RoleName:  roleName,
+		}
+		_, err = w.IamClient.AttachRolePolicy(rolePolicy)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == iam.ErrCodeEntityAlreadyExistsException {
+					return createRoleOutput.Role.Arn, nil
+				}
+				return nil, aerr
+			}
+			return nil, err
+		}
+		return createRoleOutput.Role.Arn, nil
+	}
+}
+func (w *AwsFargateWorker) CreateProfile(arn *string) error {
+	input := &eks.CreateFargateProfileInput{
+		ClusterName:         w.ClusterName,
+		FargateProfileName:  w.ProfileName,
+		PodExecutionRoleArn: arn,
+		Selectors:           w.Selectors,
+		Subnets:             w.Subnets,
+		Tags:                w.Tags,
+	}
+	// Lets see if the profile exists.  Return now if it does.
+	_, err := w.DescribeProfile()
+	if err == nil {
+		//profile exists
+		return nil
+	}
+
+	for i := 0; i < w.RetryLimit && err != nil; i++ {
+		log.Infof("CreateProfile - %d try - creating profile for cluster: %s, name: %s, arn: %s", i, *w.ClusterName, *w.ProfileName, *arn)
+		time.Sleep(time.Duration(i*500) * time.Millisecond)
+		_, err = w.EksClient.CreateFargateProfile(input)
+	}
+	return err
```

Comment:
I put the retry in because I almost always experienced race conditions between creating the role and then using the role in the fargate profile creation.

Reply:
Can you give an example? this call is just meant to start the creation, if we need to wait for the profile to be created, we should keep requeuing until it's ready (rather than waiting inside the reconcile)

Reply:
On a create, without a user-specified Role, the provisioner first creates the default Role and then creates the Profile. More often than not, the profile creation request fails because the role does not exist, even though I just created it just milliseconds before.  If I use a sleep loop, after a few trys,  the profile create request succeeds.

Reply:
I am going to pull the role creation and the profile creation apart and manage them independently through the kube polling.

Reply:
I'll commit my current implementation where they are synchronized if we want it later.